WIP: implementing marching cubes and isosurface plotting

vnredmon opened the pull request.

vnredmon committed:
WIP: implementing marching cubes

vnredmon committed:
WIP: marching cubes mostly implemented. untested.

sritchie commented on src/emmy/mathbox/marching-cubes/marching_cubes.cljs:
```diff
@@ -0,0 +1,73 @@
+      9 [xInc (interp y (+ y ystep) v2 v32 rhs) z]
+      10 [xInc (interp y yInc v4 v64 rhs)]
+      11 [(x (interp y yInc v8 v128 rhs) zInc)]
+      "Index must be between 0 and 11" ;TODO how do I throw errors in clojure?
```
`(throw (js/Error. ""))` or `(throw (ex-info "message" {:k "v"}))`

vnredmon committed:
add in triangle tables

vnredmon committed:
start on plotting stuff

vnredmon committed:
rename directory, debug namespace loading (loads!)

next up testing the marching cubes implementation for correctness

vnredmon committed:
Seems to work! gives values close to the rhs.

...for (TestLhs 1). But there are some points with only 2 entries. huh.
Need to set up tests.

vnredmon committed:
defs -> let

vnredmon committed:
Marching cubes repl test.

vnredmon committed:
WIP: Stab at plotting utilities.

vnredmon commented on src/emmy/mathbox/marching_cubes/marching_cubes.cljs:
```diff
@@ -0,0 +1,103 @@
+  (reduce +
+          (map (fn [a] (Math/pow a 2)) [x y z])))
+
+(do
```
This block gives me a vector with a bunch of points who's magnitude is pretty close to 1. 

I can add some tests as desired (probably would be good for me to do to learn clojure's testing facilities, but I can also do that independent of implementing this).

sritchie replied:
Yes please, we definitely want tests.

vnredmon replied:
Will add. 👍

vnredmon commented on src/emmy/mathbox/plot.clj:
```diff
@@ -727,3 +727,10 @@
         (ev/fragment scene))))
+(defn implicit-surface
+  [opts]
+  (let [[f-bind opts] (mc/compile-3d opts :f 3)]
```
@sritchie -- I'm getting caught up in this function. I know that `mc/compile-3d` is not what I want. But I'm murky on what's going on here -- these look like functions that take the clojure functions, turn them into clojurescript functions and return a fragment that has all the necessary details for a mathbox plot. 


I _think_ I need to implement a `compile-implicit` function that will take the Lhs and Rhs of the implicit function and do a bunch of the same stuff as `compile-3d`. Could you walk me through compile-3d and probably the `compile-state-fn` therein? I don't have a great high-level idea.

sritchie replied:
so what is going on here is that we have a clojure function on the clj side that we need to turn into something like `(js/Function. "input" "body")` for the cljs side.

I can definitely help with this; I think the easiest thing to do now is to just IGNORE this part and then pass in a quoted function, like `'(fn [in out] ....)` with whatever signature you need for your code to work.

Then once everything is humming we can add this part in, which will remove the need for quotes and make everything faster.

wdyt?

vnredmon replied:
Willing to try! I don't really know the tradeoffs, it seems like the development path will be speedier, and I'll probably learn something in the process.

vnredmon commented on src/emmy/mathbox/marching_cubes/marching_cubes.cljs:
```diff
@@ -0,0 +1,103 @@
+      9 [xInc (interp y (+ y yInc) v2 v32 rhs) z]
+      10 [xInc (interp y yInc v4 v64 rhs)]
+      11 [x (interp y yInc v8 v128 rhs) zInc]
+      (println (str "Index must be between 0 and 11. Given: " index)) ;TODO how do I throw errors in clojure?
```
Will update to throw the error as per the outdated comment.

vnredmon commented on src/emmy/mathbox/marching_cubes/marching_cubes.cljs:
```diff
@@ -0,0 +1,103 @@
+              (MarchingCubesCore lhs rhs x y z xStep yStep zStep)))))
+  )
+
+(comment
```
Grabbed this pattern from this talk -- https://www.youtube.com/watch?v=Qx0-pViyIDU

The pattern seems pretty useful. Any pros/cons in your view @sritchie?

sritchie replied:
Not sure what pattern you mean - do you mean, tests embedded as comments? I think we definitely want some solid tests in the test suite for this one instead of the comment block!

vnredmon replied:
nvm. Just learning patterns people use in clojure. Apparently some people like to leave useful repl functions in comment blocks at the bottom of files. Seems useful (maybe), but maybe also gets makes the file messier.

vnredmon committed:
Use quoted function in plot.clj

vnredmon committed:
Add implicit emmy notebook.
